test/test_nestedtensor.py

```diff
@@ -729,6 +826,124 @@ def test_linear(self, device, dtype):
+        )
+        # error case: cannot collapse a ragged dimension with the implicit batch dimension
+        self.assertRaisesRegex(
+            RuntimeError,
+            r"only a regular dimension can be collapsed with the implicit batch dimension",
+            lambda: nt.reshape(8, -1)
+        )
+        # error case: invalid collapsed size
+        nt_vector = torch.nested_tensor([torch.tensor([1.0, 2.0]), torch.tensor([3.0, 4.0])])
+        self.assertRaisesRegex(
+            RuntimeError,
+            r"shape '\[.*\]' is invalid for implicit batch size [0-9]+: "
+            r"if implicit batch size < proposed_shape\[0\], then "
+            r"implicit batch size \* next size == proposed_shape\[0\] must hold",
+            lambda: nt_vector.reshape(8)
+        )
+        # error case: no proposed dimension for splitting the implicit batch dimension
+        self.assertRaisesRegex(
+            RuntimeError,
+            r"shape '\[.*\]' is invalid for implicit batch size [0-9]+: "
+            r"if implicit batch size > proposed_shape\[0\], then "
+            r"implicit batch size == proposed_shape\[0\] \* proposed_shape\[1\] must hold",
+            lambda: nt.reshape(2)
+        )
+        # error case: proposed splitting does not multiply to the implicit batch size
+        self.assertRaisesRegex(
+            RuntimeError,
+            r"shape '\[.*\]' is invalid for implicit batch size [0-9]+: "
+            r"if implicit batch size > proposed_shape\[0\], then "
+            r"implicit batch size == proposed_shape\[0\] \* proposed_shape\[1\] must hold",
+            lambda: nt.reshape(2, 3, -1)
+        )
+        # error case: cannot split a new dimension from the implicit batch dimension
+        #             because the underlying tensors cannot be stacked
+        self.assertRaisesRegex(
+            RuntimeError,
+            r"cannot stack underlying tensors because they have different sizes",
+            lambda: nt.reshape(2, 2, -1, -1)
+        )
+        # TODO: for now there seems to be no way to trigger
+        #       "cannot stack underlying tensors" error
+        #       from different strides or uneven offsets
+        #       will have to test those error messages later
+        # normal case
+        x0 = torch.randn((2, 20), device=device, dtype=dtype)
+        x1 = torch.randn((3, 20), device=device, dtype=dtype)
+        nt = torch.nested_tensor([x0, x1])
+        pt = nt.to_padded_tensor(0.0)
+        # (2, 20) -> (2, 4, 5) -> (4, 2, 5) -> (2, 5)
+        # (3, 20) -> (3, 4, 5) -> (4, 3, 5) -> (2, 5)
+        #                                      (2, 5)
+        #                                      (2, 5)
+        #                                      (3, 5)
+        #                                      (3, 5)
+        #                                      (3, 5)
+        #                                      (3, 5)
+        ntmh = nt.reshape(2, -1, 4, 5).transpose(1, 2).reshape(8, -1, 5)
```

Comment:
`reshape(8, 3, 5)` and `reshape(8, -1, -1)` should work as well - might be worth testing

i.e. the "no inherit" and "all inherit" cases when collapsing or splitting the implicit batch dim

Reply:
a couple other cases it may be good to cover:
* (2, 20) and (3, 20) -> (2, 5, 2, 2) and (3, 5, 2, 2); i.e. splitting multiple dims at once
* collapsing multiple dims at once

It's fine if these aren't allowed right now, but it'd be good to have nice error messages for them

Reply:
> a couple other cases it may be good to cover:
> 
> * (2, 20) and (3, 20) -> (2, 5, 2, 2) and (3, 5, 2, 2); i.e. splitting multiple dims at once
> * collapsing multiple dims at once
> 
> It's fine if these aren't allowed right now, but it'd be good to have nice error messages for them

These cases are errored out and tested above in `error case: invalid collapsed size` and `error case: proposed splitting does not multiply to the implicit batch size`